cmd/api/auth/authz.go

```diff
@@ -0,0 +1,51 @@
+// Copyright KubeArchive Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package auth
+
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	authzv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientAuthzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"net/http"
+)
+
+func RBACAuthorization(sari clientAuthzv1.SubjectAccessReviewInterface) gin.HandlerFunc {
+	// FIXME Hardcoded credentials should be extracted from context when authN layer is in place
+	user := "user"
+	groups := []string{"system-authenticated"}
+
+	return func(c *gin.Context) {
+		group := c.Param("group")
+		version := c.Param("version")
+		// FIXME Resource type should be extracted from context instead from the plural
+		pluralResourceType := c.Param("resourceType")
+		namespace := c.Param("namespace")
+
+		sar, err := sari.Create(c, &authzv1.SubjectAccessReview{
+			Spec: authzv1.SubjectAccessReviewSpec{
+				User:   user,
+				Groups: groups,
+				ResourceAttributes: &authzv1.ResourceAttributes{
+					Namespace: namespace,
+					Group:     fmt.Sprintf("%s/%s", group, version),
+					Resource:  pluralResourceType[:len(pluralResourceType)-1],
+					Verb:      "get",
+				},
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"message": "Unexpected error on SAR"})
+			c.Abort()
+			return
+		}
+		if !sar.Status.Allowed {
```

Comment:
Shouldn't we wait in between?

Reply:
What do you mean? AFAIU, once we've got the SAR response we can check the status response (also, in the tests this works)

Reply:
For reference and as I am still in the process of better understanding how go works regarding with routines and contexts, I analyzed the execution thread. AFAIU it is:

1. RBACAuthorization function is called with an implementation of [SubjectAccessReviewInterface](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/authorization/v1/subjectaccessreview.go#L37)
2. SubjectAccessReviewInterface is implemented by [subjectAccessReviews](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/authorization/v1/subjectaccessreview.go#L43) for the main execution.
3. authz middleware handler calls [Create](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/authorization/v1/subjectaccessreview.go#L55) of subjectaccessreview
4. In Create [Request](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L99) is instantiated and from there, [Do](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L1200) method is called.
5. Do calls the private method [request](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L1104). This is the method in charge of performing the request, handling the retries and timeout, and canceling the context in case it's needed.

Reply:
Ok, if its a sync process from our perspective, that is fine.